src/analyze/analyze.c

```diff
@@ -2145,6 +2175,60 @@ static int do_condition(int argc, char *argv[], void *userdata) {
 }
 
 static int do_verify(int argc, char *argv[], void *userdata) {
+        int r, i = 1;
+        char **filename;
+        char *orig, *base;
+        _cleanup_free_ char *alias = NULL;
+        _cleanup_(rmrdir_and_freep) char *tempdir = NULL;
+        bool is_alias, setup_alias = false;
+
+        STRV_FOREACH(filename, strv_skip(argv, 1))
+                if (strchr(*filename, ':')) {
+                        setup_alias = true;
+                        break;
+                }
+
+        if (setup_alias) {
+                char bid[SD_ID128_STRING_MAX];
+                sd_id128_t boot_id;
+
+                r = sd_id128_get_boot(&boot_id);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to obtain boot ID");
+
+                tempdir = strjoin("/tmp/systemd-analyze-", sd_id128_to_string(boot_id, bid), "-XXXXXX");
+                if (!tempdir)
+                        return log_oom();
+
+                tempdir = mkdtemp(tempdir);
+                if (!tempdir)
+                        return log_error_errno(errno, "Failed to create temporary directory");
+        }
+
+        STRV_FOREACH(filename, strv_skip(argv, 1)) {
```

Comment:
I'm not sure I follow here - the tmpdir is set up only if there's an alias, which is good, but then symlinks are created unconditionally? But tempdir won't exist?

We don't want to create symlinks unnecessarily I think. This would probably look better by doing something like:

```
r = extract_many_words(..., orig, alias, NULL)
if (r == 2)
   symlink...
```

etc

Reply:
I see that tempdir won't exist sometimes - that clearly needs to be fixed. And I agree that we shouldn't symlink if there are *no* aliases. But the intention here was to symlink everything if we symlink anything, in order to keep it all under one path.

If we don't do this, passing e.g., two valid aliases and one invalid file without an alias, such as
```
systemd-analyze verify /tmp/source:alias0.service /tmp/source2:alias1.service /tmp/source3
```
will cause *all* of the checks to fail, with `systemd-anaylze` complaining that `alias0.service` and `alias1.service` don't exist. I traced this to `generate_var` in `verify_units` but I'm not sure what the right approach is.

Reply:
But that's pretty much expected. What should work is `systemd-analyze verify /tmp/source:alias0.service /tmp/source2:alias1.service /tmp/source3-valid.service` - without symlinking the last one

Reply:
Just checked, this doesn't work either. The aliases will run as expected but you'll get unit not found for `source3-valid.service`.

Reply:
Actually I get very strange results here. This works with one test unit file I have but not another, even though they are identical in content and in the same directory.

Reply:
Hmm, I take that back, one of them is a symlink.
So we have:
```
$ ls -la /tmp
-rw-r--r--  1 albert albert    104 Jul  1 16:23 source
-rw-rw-r--  1 albert albert      0 Jul  1 16:32 source2
-rw-rw-r--  1 albert albert    104 Jul  1 16:22 test.service
lrwxrwxrwx  1 albert albert      6 Jul  2 08:33 valid-link.service -> source

$ systemd-analyze verify /tmp/source:alias.service 
# Works

$ systemd-analyze verify /tmp/source:alias.service /tmp/source2
Unit alias.service not found.
Failed to prepare filename /tmp/source2: Invalid argument

$ systemd-analyze verify /tmp/source:alias.service /tmp/test.service
Unit alias.service not found.
# test.service works though

$ systemd-analyze verify /tmp/source:alias.service /tmp/valid-link.service
Unit alias.service not found.
Unit valid-link.service not found.
```

Reply:
I think the reason those fail is because you are removing the full path from the variable - it's necessary to find the units if they are not in the canonical directories

Reply:
These don't fail with the current implementation. Up to you whether it's too hacky to symlink everything―if so I'd appreciate some guidance on how to handle this because I dug through the source with `gdb` for hours and couldn't find the root of this issue.

Reply:
Hard linking only the aliased files works fine in all cases - systemd doesn't get upset about invalid filenames at link destinations, failures stay contained within their relevant unit, everything gets cleaned up at the end

Reply:
hard linking only works if the files are on the same filesystem - try to have the original on a different FS, and it will fail

Reply:
What do you think about copying?

Reply:
That should be fine